tests/unit/conftest.py

```diff
@@ -0,0 +1,134 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""pytest fixtures for the unit test."""
+
+import contextlib
+import typing
+import unittest.mock
+
+import ops
+import pytest
+from ops.model import BlockedStatus, Container
+from ops.testing import Harness
+
+from tests.unit import helpers
+from tests.unit._patched_charm import DiscourseCharm, pgsql_patch
+
+BLOCKED_STATUS = BlockedStatus.name  # type: ignore
+
+DATABASE_NAME = "discourse"
+
+
+@pytest.fixture(name="start_harness", autouse=True)
+def start_harness_fixture():  # -> typing.Generator[Harness, None, None]:
+    """Ops testing framework harness fixture."""
+
+    def start_harness(
+        *,
+        with_postgres: bool = True,
+        with_redis: bool = True,
+        with_ingress: bool = False,
+        with_config: typing.Optional[typing.Dict[str, typing.Any]] = None,
+    ):
+        """Start a harness discourse charm.
+
+        This is also a workaround for the fact that Harness
+        doesn't reinitialize the charm as expected.
+        Ref: https://github.com/canonical/operator/issues/736
+
+        Args:
+            - with_postgres: should a postgres relation be added
+            - with_redis: should a redis relation be added
+        """
+        harness = Harness(DiscourseCharm)
+        harness.disable_hooks()
+        # Protected access is needed to instantiate harness
+        # pylint: disable=protected-access
+        harness._framework = ops.framework.Framework(
+            # pylint: disable=protected-access
+            harness._storage,
+            harness._charm_dir,
+            harness._meta,
+            harness._model,
+        )
+        if with_postgres:
+            _add_postgres_relation(harness)
+        harness.enable_hooks()
+        harness.begin_with_initial_hooks()
+        if with_redis:
+            _add_redis_relation(harness)
+            with helpers.patch_exec():
+                harness.framework.reemit()
+        if with_ingress:
+            _add_ingress_relation(harness)
+
+        with helpers.patch_exec(), _patch_setup_completed():
+            charm: DiscourseCharm = typing.cast(DiscourseCharm, harness.charm)
+            charm._set_setup_completed()
+
+        pgsql_patch.start()
+
+        if with_config is not None:
+            harness.update_config(with_config)
+            with helpers.patch_exec():
+                harness.container_pebble_ready("discourse")
+
+        return harness
+
+    return start_harness
+
+
+@contextlib.contextmanager
+def _patch_setup_completed():
```

Comment:
Don't we have a new way of doing this in tests with being able to interact with the file system of the container?

Reply:
A link to an implementation would be helpful here.

Reply:
@weiiwang01 could you help please? I think this fixture uses both for the command execution and the file system: https://github.com/canonical/flask-k8s-operator/blob/609d60e723974f46269139dd2b1539e4b65fe592/tests/unit/conftest.py#L85

Reply:
I think that you're talking about the [get_filesystem_root](https://ops.readthedocs.io/en/latest/#ops.testing.Harness.get_filesystem_root) method.  
I tried to use it that way:
```
    def _set_setup_completed():
        root = harness.get_filesystem_root("discourse")
        (root / SETUP_COMPLETED_FLAG_FILE[1:]).parent.mkdir(parents=True, exist_ok=True)
        (root / SETUP_COMPLETED_FLAG_FILE[1:]).write_text("")
    harness.charm._set_setup_completed = _set_setup_completed
```

Before realizing that the tests didn't rely on it. So I removed that.